### PR TITLE
Patch generate files for system protobuf

### LINKED ISSUE #42939: Failed to build with CUDA 11 on Windows when system protobuf is used:  Ambiguous operator overload.
## Bug

Get the following error when building with CUDA.
```
D:/roaster-scratch/pytorch/build\caffe2/proto/caffe2.pb.h(151): error: more than one operator "+" matches these operands:
            built-in operator "arithmetic + arithmetic"
            function "c10::operator+(int, c10::BFloat16)"
            function "c10::operator+(c10::BFloat16, int)"
            function "c10::operator+(int, c10::Half)"
            function "c10::operator+(c10::Half, int)"
            operand types are: const caffe2::ExternalDataProto_SourceType + int

D:/roaster-scratch/pytorch/build\caffe2/proto/caffe2.pb.h(186): error: more than one operator "+" matches these operands:
            built-in operator "arithmetic + arithmetic"
            function "c10::operator+(int, c10::BFloat16)"
            function "c10::operator+(c10::BFloat16, int)"
            function "c10::operator+(int, c10::Half)"
            function "c10::operator+(c10::Half, int)"
            operand types are: const caffe2::TensorProto_DataType + int

D:/roaster-scratch/pytorch/build\caffe2/proto/caffe2.pb.h(211): error: more than one operator "+" matches these operands:
            built-in operator "arithmetic + arithmetic"
            function "c10::operator+(int, c10::BFloat16)"
            function "c10::operator+(c10::BFloat16, int)"
            function "c10::operator+(int, c10::Half)"
            function "c10::operator+(c10::Half, int)"
            operand types are: const caffe2::TensorProto_StorageType + int

D:/roaster-scratch/pytorch/build\caffe2/proto/caffe2.pb.h(239): error: more than one operator "+" matches these operands:
            built-in operator "arithmetic + arithmetic"
            function "c10::operator+(int, c10::BFloat16)"
            function "c10::operator+(c10::BFloat16, int)"
            function "c10::operator+(int, c10::Half)"
            function "c10::operator+(c10::Half, int)"
            operand types are: const caffe2::TensorBoundShape_DimType + int

D:/roaster-scratch/pytorch/build\caffe2/proto/caffe2.pb.h(272): error: more than one operator "+" matches these operands:
            built-in operator "arithmetic + arithmetic"
            function "c10::operator+(int, c10::BFloat16)"
            function "c10::operator+(c10::BFloat16, int)"
            function "c10::operator+(int, c10::Half)"
            function "c10::operator+(c10::Half, int)"
            operand types are: const caffe2::DeviceTypeProto + int

5 errors detected in the compilation of "D:/roaster-scratch/pytorch/caffe2/operators/utility_ops.cu".
utility_ops.cu
-- Removing D:/roaster-scratch/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/operators/./torch_cuda_generated_utility_ops.cu.obj
"C:/Program Files/CMake/bin/cmake.exe" -E remove D:/roaster-scratch/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/operators/./torch_cuda_generated_utility_ops.cu.obj
CMake Error at torch_cuda_generated_utility_ops.cu.obj.Release.cmake:281 (message):
  Error generating file
  D:/roaster-scratch/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/operators/./torch_cuda_generated_utility_ops.cu.obj


ninja: build stopped: subcommand failed.
```

Problematic lines in generated proto header:
```
constexpr int ExternalDataProto_SourceType_SourceType_ARRAYSIZE = ExternalDataProto_SourceType_SourceType_MAX + 1;
constexpr int TensorProto_DataType_DataType_ARRAYSIZE = TensorProto_DataType_DataType_MAX + 1;
constexpr int TensorProto_StorageType_StorageType_ARRAYSIZE = TensorProto_StorageType_StorageType_MAX + 1;
constexpr int TensorBoundShape_DimType_DimType_ARRAYSIZE = TensorBoundShape_DimType_DimType_MAX + 1;
constexpr int DeviceTypeProto_ARRAYSIZE = DeviceTypeProto_MAX + 1;
```
image

## Environment

 - PyTorch Version (e.g., 1.0): master
 - OS (e.g., Linux): Win Server 2019
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): cmake+ninja+msvc2019
 - CUDA/cuDNN version: 11.0/8

cc @malfet @seemethere @walterddr @ngimel @peterjc123 @maxluk @nbcsm @guyang3532 @gunandrose4u @smartcat2010 @mszhanyi

### PR DESCRIPTION
**Pull request by peterjc123**
Fixes https://github.com/pytorch/pytorch/issues/42939

### CODE REVIEW

File: cmake/ProtoBuf.cmake [hunk 1]
```diff
@@ -196,7 +196,7 @@ function(caffe2_protobuf_generate_cpp_py srcs_var hdrs_var python_var)
 
         # If we remove all reference to these pb.h files from external
         # libraries and binaries this rewrite can be removed.
-        COMMAND ${CMAKE_COMMAND} -DFILENAME=${CMAKE_CURRENT_BINARY_DIR}/${fil_we}.pb.h -DNAMESPACES=caffe\;caffe2\;onnx\;torch -P ${PROJECT_SOURCE_DIR}/cmake/ProtoBufPatch.cmake
+        COMMAND ${CMAKE_COMMAND} -DFILENAME=${CMAKE_CURRENT_BINARY_DIR}/${fil_we}.pb.h -DNAMESPACES=caffe\;caffe2\;onnx\;torch -DSYSTEM_PROTOBUF=OFF -P ${PROJECT_SOURCE_DIR}/cmake/ProtoBufPatch.cmake
```
Comment (xkszltl) [inline cmake/ProtoBuf.cmake:199]: Suggest to unify these 2 lines to `-DSYSTEM_PROTOBUF=${CAFFE2_LINK_LOCAL_PROTOBUF}`

**Review by xkszltl — CHANGES_REQUESTED** [review]

File: cmake/ProtoBuf.cmake (hunk 1 above)
Comment (peterjc123) [reply cmake/ProtoBuf.cmake:199]: Sounds reasonable.

Comment (peterjc123) [reply cmake/ProtoBuf.cmake:199]: Wait, it should be !CAFFE2_LINK_LOCAL_PROTOBUF, right?

Comment (xkszltl) [reply cmake/ProtoBuf.cmake]: Emmm yes.....then this make sense.

Comment (xkszltl) [reply cmake/ProtoBuf.cmake]: OK the `LOCAL_PROTOBUF` approach is better.
BTW why not just pass in CAFFE2_LINK_LOCAL_PROTOBUF. grep experience is better that way

**Review by ezyang — APPROVED** [review]
easier to review when whitespace changes were ignored

**Comment by ezyang** [comment]
The test failures might be legit, they've been sticky past merge to master.

**Review by ezyang — CHANGES_REQUESTED** [review]
test failures

**Comment by peterjc123** [comment]
@ezyang The test failures don't seem to be related now.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**